### PR TITLE
MYST-703 Refactor CLI flag bootstraping

### PR DESCRIPTION
**Pull request by Waldz**

### CODE REVIEW

File: cmd/di.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		identityLoader,
+		di.SignerFactory,
+		di.IdentityRegistry,
+		di.MysteriumClient,
+		di.IPResolver,
+		di.LocationResolver,
+	)
+}
+
+// function decides on etwork definition combined from testnet/localnet flags and possible overrides
```
Comment (soffokl) [inline cmd/di.go]: etwork -> network

File: openvpn/core/options_node.go [hunk 2]
```diff
@@ -27,14 +27,19 @@ import (
 	"strconv"
 	"syscall"
 
-	log "github.com/cihub/seelog"
+	"github.com/ethereum/go-ethereum/log"
```
Comment (soffokl) [inline openvpn/core/options_node.go]: Looks like wrong logging package

File: identity/registry/registration.go [hunk 3]
```diff
@@ -18,15 +18,15 @@
...
 	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	"github.com/ethereum/go-ethereum/accounts/keystore"
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/mysteriumnetwork/payments/registry"
+	"github.com/mysteriumnetwork/payments/registry/generated"
 )
 
-// IdentityRegistry checks whenever given identity is registered
-type IdentityRegistry interface {
+// Registry checks whenever given identity is registered
```
Comment (tadovas) [inline identity/registry/registration.go]: Registry of what? It either should be in identity package or old name should be returned. Current import ant usage doesn't sound: import "...../registry" registry.Registry

File: openvpn/core/flags_node.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+/*
```
Comment (tadovas) [inline openvpn/core/flags_node.go]: openvpn/core ? These flags are not related to core of openvpn. We tried so hard to move out all code not directly related to openvpn functionality itself.

File: openvpn/core/options_node.go (hunk 2 above)
Comment (Waldz) [reply openvpn/core/options_node.go]: Fixed

File: identity/registry/registration.go (hunk 3 above)
Comment (Waldz) [reply identity/registry/registration.go]: It should belong to Openvpn Adapter in future, we will mote out later

File: openvpn/core/flags_node.go (hunk 4 above)
Comment (Waldz) [reply openvpn/core/flags_node.go]: It should belong to Openvpn Adapter in future, we will mote out later

File: identity/registry/registration.go (hunk 3 above)
Comment (Waldz) [reply identity/registry/registration.go]: This is not package `registry`, but rather parent/child package `identity/registry`. This is why I allias those package then importing:
```
identity_loading "github.com/mysteriumnetwork/node/identity/loading"
identity_registry "github.com/mysteriumnetwork/node/identity/registry"
``` 

So I thought there is not point to repeat name `identity_loading.IdentityRegistry`. But aliasing each time is easy to forget, so I dont know perfect pattern in Golang :/

**Comment by Waldz** [comment]
@tadovas @soffokl All comments fixed

File: cmd/di.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+func (di *Dependencies) bootstrapIdentity(directories node.DirectoryOptions) {
+	di.Keystore = identity.NewKeystoreFilesystem(directories.Keystore)
+	di.IdentityManager = identity.NewIdentityManager(di.Keystore)
+	di.SignerFactory = func(id identity.Identity) identity.Signer {
+		return identity.NewSigner(di.Keystore, id)
+	}
+}
+
+func (di *Dependencies) bootstrapLocation(options node.LocationOptions, configDirectory string) {
```
[the remark this answers is not in the document]
Comment (Waldz) [reply cmd/di.go]: I bootstrap all location configs here, it's like "location" module

File: core/node/options.go [hunk 6]
```diff
@@ -17,16 +17,16 @@
...
 	Directories DirectoryOptions
 
 	TequilapiAddress string
 	TequilapiPort    int
 
-	OpenvpnBinary    string
-	IpifyUrl         string
-	LocationDatabase string
-
+	Openvpn  openvpn_core.NodeOptions
```
[the remark this answers is not in the document]
Comment (Waldz) [reply core/node/options.go]: Working on that in another PR

File: cmd/flags_location.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
+/*
```
[the remark this answers is not in the document]
Comment (Waldz) [reply cmd/flags_location.go]: File name is not so important `flags_network.go` contains logic related to network flags only: flag constants, registering and parsing

File: core/node/options_location.go [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2017 The "MysteriumNetwork/node" Authors.
+ * Copyright (C) 2018 The "MysteriumNetwork/node" Authors.
```
[the remark this answers is not in the document]
Comment (Waldz) [reply core/node/options_location.go]: I splited file `options.go` to smaller pieces, this is why
screen shot 2018-09-07 at 14 07 07

[the remark this answers is not in the document]
Comment (tadovas) [reply core/node/options_location.go]: OK. Lets keep that

File: core/node/options_network.go [hunk 9]
```diff
@@ -17,11 +17,6 @@
```
[the remark this answers is not in the document]
Comment (Waldz) [reply core/node/options_network.go]: Made consistent

File: identity/loading/loading.go [hunk 10]
```diff
@@ -15,22 +15,24 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-package identity
+package loading
```
[the remark this answers is not in the document]
Comment (Waldz) [reply identity/loading/loading.go]: Renamed to `identity/selector`

File: cmd/di.go [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+	if options.DiscoveryAPIAddress != metadata.DefaultNetwork.DiscoveryAPIAddress {
+		network.DiscoveryAPIAddress = options.DiscoveryAPIAddress
+	}
+
+	if options.BrokerAddress != metadata.DefaultNetwork.BrokerAddress {
+		network.BrokerAddress = options.BrokerAddress
+	}
+
+	normalizedAddress := common.HexToAddress(options.EtherPaymentsAddress)
+	if normalizedAddress.String() != metadata.DefaultNetwork.PaymentsContractAddress.String() {
```
Comment (zolia) [inline cmd/di.go]: Do we really need to convert both types to String? Struct should be comparable here

File: cmd/di.go [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+	di.IdentityManager = identity.NewIdentityManager(di.Keystore)
+	di.SignerFactory = func(id identity.Identity) identity.Signer {
+		return identity.NewSigner(di.Keystore, id)
+	}
+}
+
+func (di *Dependencies) bootstrapLocation(options node.OptionsLocation, configDirectory string) {
+	di.IPResolver = ip.NewResolver(options.IpifyUrl)
+
+	switch {
```
Comment (zolia) [inline cmd/di.go]: Two options here, simple if / else should be enough.

File: core/ip/rest_resolver.go [hunk 13]
```diff
@@ -52,13 +52,17 @@ func NewResolverWithTimeout(ipifyUrl string, timeout time.Duration) Resolver {
 	}
 }
 
+type ipResponse struct {
```
Comment (zolia) [inline core/ip/rest_resolver.go]: imho bad name here, should be resultingIP or returnedIP

File: cmd/di.go [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+	log.Info("Using Contract at address:", network.PaymentsContractAddress.String())
+	if di.IdentityRegistry, err = identity_registry.NewIdentityRegistry(di.EtherClient, network.PaymentsContractAddress); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (di *Dependencies) bootstrapIdentity(directories node.DirectoryOptions) {
```
[the remark this answers is not in the document]
Comment (Waldz) [reply cmd/di.go]: Rebamed all functions

File: core/node/node.go [hunk 15]
```diff
@@ -45,42 +44,33 @@ import (
 )
 
 // NewNode function creates new Mysterium node by given options
-func NewNode(options Options) *Node {
-	networkDefinition := GetNetworkDefinition(options.NetworkOptions)
-	mysteriumClient := server.NewClient(networkDefinition.DiscoveryAPIAddress)
-
+func NewNode(
```
Comment (zolia) [inline core/node/node.go]: We could use NodeOption or NodeFlags structure here, not multi parameter factory.. as per: https://hackmongo.com/page/golang-antipatterns/#runaway-arguments

File: cmd/di.go (hunk 11 above)
Comment (Waldz) [reply cmd/di.go]: Remade

File: core/ip/rest_resolver.go (hunk 13 above)
Comment (Waldz) [reply core/ip/rest_resolver.go]: Not my scope. And all structures in tequilapi endpoints has `*Response` and `*Request` naming

File: identity/registry/registry_contract.go [hunk 16]
```diff
@@ -18,51 +18,35 @@
...
-// NewRegistrationDataProvider creates registration data provider backed up by identity which is managed by keystore
-func NewRegistrationDataProvider(ks *keystore.KeyStore) RegistrationDataProvider {
-	return &keystoreRegistrationDataProvider{
-		ks: ks,
-	}
-}
-
 // NewIdentityRegistry creates identity registry service which uses blockchain for information
-func NewIdentityRegistry(contractCaller bind.ContractCaller, registryAddress common.Address) (IdentityRegistry, error) {
+func NewIdentityRegistry(contractCaller bind.ContractCaller, registryAddress common.Address) (*contractRegistry, error) {
```
Comment (zolia) [inline identity/registry/registry_contract.go]: I would leave IdentityRegistry as is, since the only way identities can be registered is through contract.
Factory method also is called NewIdentityRegistry not NewContractRegistry.

File: openvpn/core/options_node.go [hunk 17]
```diff
@@ -30,11 +30,16 @@ import (
 	log "github.com/cihub/seelog"
 )
 
+// NodeOptions describes possible parameters of Openvpn configuration
+type NodeOptions struct {
+	Binary string
```
Comment (zolia) [inline openvpn/core/options_node.go]: Would rename it to BinaryPath

File: core/node/node.go (hunk 15 above)
Comment (Waldz) [reply core/node/node.go]: Yeah, that's what i'm moving towards. But not so easy while we have 3 executables. Would like to do it in upcoming PR

File: identity/registry/registry_contract.go (hunk 16 above)
Comment (Waldz) [reply identity/registry/registry_contract.go]: We dont return interface type, but type of implementation. This is why I renamed

File: openvpn/core/options_node.go (hunk 17 above)
Comment (Waldz) [reply openvpn/core/options_node.go]: Renamed

File: identity/registry/registry_contract.go (hunk 16 above)
Comment (Waldz) [reply identity/registry/registry_contract.go]: Renamed to `NewIdentityRegistryContract()`

**Comment by Waldz** [comment]
@tadovas @zolia all comments fixed
